### ISSUE TITLE
Error: No such file or directory - cf

### ISSUE THREAD

**Issue by easuncion-solace** [cloudfoundry/homebrew-tap#48]
Ran "brew install cf-cli", encountered "Error: No such file or directory - cf" message.

**Comment by XenoPhex** [comment]
by chance - do you have `bash-completion` installed?

**Comment by easuncion-solace** [comment]
I am using zsh so this maybe the issue.

**Comment by XenoPhex** [comment]
our formula might be assuming the existence of bash-completion, we can adjust based off of that. We're testing what happens in a few different cases. Hopefully we'll have a fix in the next few hours!

**Comment by ruurd** [comment]
I use bash, same problem.

**Comment by ruurd** [comment]
I have a look at the bosh_cli recipe that does install.bin bosh_cli_executable => bosh

**Comment by XenoPhex** [comment]
can y'all provide us the `brew config` details? I just tried running things against a fresh install of homebrew and it worked. So there must be something non-bash related that's causing an issue.

**Comment by XenoPhex** [comment]
@ruurd I believe that's only necessary if the binary has a different name post-install than what it was in the package download.

**Comment by ruurd** [comment]
Well, that begs the question if there is a cf in the downloaded package, amirite? Ran brew doctor and fixed any brew problems, error persists.

**Comment by ruurd** [comment]
$ brew config
HOMEBREW_VERSION: 1.7.1-204-g8e703e3
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: 8e703e30ba53f8089686b0fc90f87803b4fd5d90
Last commit: 3 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 4257cb76a87f331f3bee7f91dfb9754bb00fd826
Core tap last commit: 35 minutes ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_DEV_CMD_RUN: 1
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_VISUAL: nano
CPU: octa-core 64-bit haswell
Homebrew Ruby: 2.3.7 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/bin/ruby
Clang: 9.1 build 902
Git: 2.18.0 => /usr/local/bin/git
Curl: 7.54.0 => /usr/bin/curl
Java: 11, 10.0.1, 1.8.0_181, 1.7.0_80
macOS: 10.13.6-x86_64
CLT: 10.0.0.0.1.1527767617
CLT headers: N/A
Xcode: 9.4.1
XQuartz: 2.7.11 => /opt/X11

**Comment by XenoPhex** [comment]
The first thing I confirmed is that the correct binary exists in the [url](https://github.com/cloudfoundry/homebrew-tap/blob/master/cf-cli.rb#L9) provided by formula.

**Comment by anwarchk** [comment]
Homebrew config:

```ORIGIN: https://github.com/Homebrew/brew
HEAD: 8e703e30ba53f8089686b0fc90f87803b4fd5d90
Last commit: 2 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: a0f27d7db954b326fac40040b8509b3bed8b7e28
Core tap last commit: 76 minutes ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_DEV_CMD_RUN: 1
HOMEBREW_NO_ANALYTICS_THIS_RUN: 1
CPU: octa-core 64-bit kabylake
Homebrew Ruby: 2.3.7 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/bin/ruby
Clang: 9.1 build 902
Git: 2.18.0 => /usr/local/bin/git
Curl: 7.54.0 => /usr/bin/curl
Java: 1.8.0_172, 1.8.0_162, 1.8.0_152, 1.8.0_141, 1.8.0_131, 1.8.0_121, 1.8.0_112, 1.8.0_102, 1.8.0_92, 1.8.0_77, 1.7.0_79
macOS: 10.13.6-x86_64
CLT: 9.4.1.0.1.1528165917
```

**Comment by XenoPhex** [comment]
BTW - Thanks y'all for being so prompt!

**Comment by anwarchk** [comment]
I do have `bash-completion` installed

**Comment by ruurd** [comment]
I somehow can't reproduce the sha256 of the package. But can be that auto unpacking plays havoc here.

**Comment by XenoPhex** [comment]
by chance - how did y'all install homebrew? via the standard:
```bash
/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
```
or via the alternative install:
```bash
mkdir homebrew && curl -L https://github.com/Homebrew/brew/tarball/master | tar xz --strip 1 -C homebrew
```

Because the one thing we notice in both your configs is that `brew` is on `HEAD` of homebrew - not the latest tag (which has a sha1 of `d6a245c05110c240e680da594c2cc6038998b82e`, Tag: `1.7.1`).

We're still looking into it (as this may be an issue for a future version of homebrew) - but can y'all try with the latest _tagged_ version and see if you still see an issue?

**Comment by ruurd** [comment]
Funny, in the cache after downloading it, there is no extension for some reason. If I look at it with file then it is a tar-gzip archive. If I look at the cache entries for bosh they are just executables.

**Comment by XenoPhex** [comment]
After running git bisect on the [official homebrew repo](https://github.com/Homebrew/brew) - we've determined that [this commit](https://github.com/Homebrew/brew/commit/4c80f827) will break us in the future. We're going to try and contact the official maintainers to figure out a good solution going forward.

For now - we'd recommend only running the officially tagged versions of homebrew - not their unreleased `master` branch. Hopefully we'll have this resolved before releasing their next version.

**Comment by anwarchk** [comment]
For any brew command, what brew does is an `auto update`, which always checks out the master version. How does one prevent it from auto updating it so that I can stay on the tag `1.7.1` ? @XenoPhex @ruurd ?

**Comment by anwarchk** [comment]
nm, found this `HOMEBREW_NO_AUTO_UPDATE=1`

**Comment by XenoPhex** [comment]
@anwarchk you can also do a `git checkout stable` instead of being on `master` in your `/usr/local/Homebrew` directory.

**Comment by ruurd** [comment]
@XenoPhex how does that relate to the fact that the cached download does not have a file extension?

**Comment by tjvman** [comment]
@ruurd  Homebrew on `master` searches by extension only when looking for the downloaded files for a given formula, but it's not downloading the tarball to a file with the correct extension (as you have pointed out). There's some more detail in the bug linked above.

**Comment by ocean90** [comment]
I was using the standard installation but still had the `homebrew.devcmdrun` config enabled. You can run`git -C $(brew --repo) config --unset homebrew.devcmdrun && brew update --force` to get back to the stable branch/tag.

**Comment by martys34** [comment]
Also faced a similar issue: tried to run `brew install cf-cli` and received the error `Error: No such file or directory - cf`.
I tried to uninstall --force, prune, cleanup and tried installing it again but it would still fail at: 

```
cp -p /Users/*/Library/Caches/Homebrew/cf-cli--6.38.0 /private/tmp/cf-cli-20180820-53542-c5fkwl/stable
```
I ended up using the OSX Installer which worked.

**Comment by voelzmo** [comment]
Found this to be broken with the latest release 1.7.2 as well. Homebre release 1.7.1 seems to be the last correct one.
For reference, here is what I did to install the newest cf-cli over homebrew for now:
```
cd /usr/local/Homebrew
git checkout 1.7.1
rm -rf /Users/d058546/Library/Caches/Homebrew/cf-cli--6.38.0
HOMEBREW_NO_AUTO_UPDATE=0 brew upgrade cf-cli
```

**Comment by dtfiedler** [comment]
This worked! Thanks @voelzmo

**Comment by XenoPhex** [comment]
We were hoping that the homebrew team was going to fix the issue we filed (Homebrew/brew#4648) before releasing but that turned out to not be the case. We're going to try something to see if we can mitigate this problem in a different way.

**Comment by Jackman3005** [comment]
Thanks @voelzmo that led me to the solution. For me it was:
```bash
cd /usr/local/Homebrew
git checkout 1.7.1
HOMEBREW_NO_AUTO_UPDATE=0 brew tap cloudfoundry/tap
HOMEBREW_NO_AUTO_UPDATE=0 brew install cf-cli
```

For some reason it wanted me to do a new tap? We didn't have cf-cli installed at all yet on the machine.

**Comment by XenoPhex** [comment]
I JUST updated the formula to use a url with a _more_ valid filename. This should have fixed the issue. (Be sure to do a `brew update`)

I'm going to leave this issue open until tomorrow on the off chance people are still having issue after the fix.

**Comment by voelzmo** [comment]
@XenoPhex Thanks! I can confirm this is fixed now with homebrew 1.7.2 and most recent formula for cf-cli 6.28.0

**Comment by XenoPhex** [comment]
Everything looks good now. Please feel free to reopen if you're experiencing more issues.
